File: core/src/main/java/mifos/org/mobilewallet/core/domain/usecase/FetchAccountTransactions.java
package mifos.org.mobilewallet.core.domain.usecase;

import java.util.List;

import javax.inject.Inject;

import mifos.org.mobilewallet.core.base.UseCase;
import mifos.org.mobilewallet.core.data.fineract.repository.FineractRepository;
import mifos.org.mobilewallet.core.data.local.LocalRepository;
import mifos.org.mobilewallet.core.domain.model.Transaction;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by naman on 11/7/17.
 */

public class FetchAccountTransactions extends UseCase<FetchAccountTransactions.RequestValues,
        FetchAccountTransactions.ResponseValue> {

    private final LocalRepository localRepository;
    private final FineractRepository fineractRepository;

    @Inject
    public FetchAccountTransactions(LocalRepository localRepository, FineractRepository fineractRepository) {
        this.localRepository = localRepository;
        this.fineractRepository = fineractRepository;
    }


    @Override
    protected void executeUseCase(final FetchAccountTransactions.RequestValues requestValues) {

        fineractRepository.getAccountTransactions(requestValues.accountId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<List<Transaction>>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        getUseCaseCallback().onError("Error fetching remote account transactions");
                    }

                    @Override
                    public void onNext(List<Transaction> transactions) {
                        getUseCaseCallback().onSuccess(new ResponseValue(transactions));
                    }
                });



    }

    public static final class RequestValues implements UseCase.RequestValues {

        private long accountId;

        public RequestValues(long accountId) {
            this.accountId = accountId;
        }

        public void setAccountId(long accountId) {
            this.accountId = accountId;
        }
    }

    public static final class ResponseValue implements UseCase.ResponseValue {

        private List<Transaction> transactions;

        public ResponseValue(List<Transaction> transactions) {
            this.transactions = transactions;
        }

        public List<Transaction> getTransactions() {
            return transactions;
        }
    }

}

File: core/data/src/main/java/org/mifospay/core/data/domain/usecase/account/FetchAccountTransactions.kt
package org.mifospay.core.data.domain.usecase.account

import com.mifospay.core.model.domain.Transaction
import com.mifospay.core.model.entity.accounts.savings.SavingsWithAssociations
import org.mifospay.core.data.base.UseCase
import org.mifospay.core.data.fineract.entity.mapper.TransactionMapper
import org.mifospay.core.data.fineract.repository.FineractRepository
import org.mifospay.core.data.util.Constants
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created by naman on 11/7/17.
 */
class FetchAccountTransactions @Inject constructor(
    private val fineractRepository: FineractRepository,
    private val transactionMapper: TransactionMapper
) : UseCase<FetchAccountTransactions.RequestValues, FetchAccountTransactions.ResponseValue?>() {

    protected override fun executeUseCase(requestValues: RequestValues) {
        fineractRepository.getSelfAccountTransactions(requestValues.accountId)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : Subscriber<SavingsWithAssociations>() {
                override fun onCompleted() {}
                override fun onError(e: Throwable) {
                    useCaseCallback.onError(
                        Constants.ERROR_FETCHING_REMOTE_ACCOUNT_TRANSACTIONS
                    )
                }

                override fun onNext(transactions: SavingsWithAssociations) {
                    useCaseCallback.onSuccess(
                        ResponseValue(
                            transactionMapper.transformTransactionList(transactions)
                        )
                    )
                }
            })
    }

    data class RequestValues(var accountId: Long) : UseCase.RequestValues
    data class ResponseValue(val transactions: List<Transaction>) : UseCase.ResponseValue
}
